<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-dark.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>CNV-Sim by NabaviLab</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>CNV-Sim</h1>
          <h2>Copy Number Variations (CNV) Simulator</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/NabaviLab/CNV-Sim/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/NabaviLab/CNV-Sim/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/NabaviLab/CNV-Sim" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">
          <h1>
<a id="copy-number-variation-simulator-cnv-sim" class="anchor" href="#copy-number-variation-simulator-cnv-sim" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Copy Number Variation Simulator (CNV-Sim)</h1>

<p>In genomics, Copy Number Variations (CNVs) is a type of structural variation in a genome where sections of the genome are duplicated or deleted. 
The number of variations (duplications/deletions) varies between individuals in the human population.</p>

<p>The Copy Number Variation Simulator (CNV-Sim) is a simulation tool that extends the functionality of existing next-generation sequencing read simulators 
to introduce copy number variations in the generated reads. The resulting reads encompass amplifications as well as deletions according to a predefined list of variant regions.</p>

<p>CNV-Sim aids testing and benchmarking tools for copy number variation detection and analysis. The tool offers two types of simulation:</p>

<ul>
<li>CNV simulation in whole genome. CNV-Sim utilizes the functionality of <a href="http://www.niehs.nih.gov/research/resources/software/biostatistics/art/">ART</a> to introduce variations in the genome.</li>
<li>CNV simulation in targeted exome. CNV-Sim utilizes the functionality of <a href="https://github.com/sak042/Wessim">Wessim</a> to introduce variations in the targets.<br>
</li>
</ul>

<h2>
<a id="use-cnv-sim" class="anchor" href="#use-cnv-sim" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Use CNV-Sim</h2>

<h3>
<a id="download-as-a-standalone-application" class="anchor" href="#download-as-a-standalone-application" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download as a standalone application</h3>

<p>Run the setup script appropriate for your operating system to install required dependencies. </p>

<p><em>Alternatively</em>, install the following dependencies on your machine (the setup script does all that for you):</p>

<ul>
<li><a href="https://www.python.org/downloads/">Python 2.7</a></li>
<li><a href="https://pypi.python.org/pypi/pip">Python pip</a></li>
<li><a href="https://github.com/pysam-developers/pysam">Pysam Module</a></li>
<li><a href="http://biopython.org/">Biopython Module</a></li>
<li><a href="http://bedtools.readthedocs.io/en/latest/">bedtools</a></li>
<li>Download <a href="http://www.niehs.nih.gov/research/resources/software/art">ART simulator</a> and put the binary file <code>art_illumina</code> in <code>cnvsim/ART</code> directory.</li>
</ul>

<p>Refer to the below <a href="#cnv-sim-options">CNV-Sim options</a> section for more details on how to use all available command-line options</p>

<h3>
<a id="download-as-a-docker-image" class="anchor" href="#download-as-a-docker-image" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download as a Docker image</h3>

<p>We prefer that you run CNV-Sim from the <a href="http://www.docker.com">Docker</a> container as it has all the dependencies installed inside yet; no need to go through any of the setup scripts included here.</p>

<p><strong>New to Docker?</strong> Read this <a href="https://www.toptal.com/devops/getting-started-with-docker-simplifying-devops">blog post</a> to understand how it works
(yet you don't need to be a Docker expert to get CNV-Sim Docker container running!).</p>

<p><strong>Install Docker:</strong> <a href="https://docs.docker.com/engine/installation/#/on-linux">Linux</a>, <a href="https://docs.docker.com/docker-for-mac/">Mac OS</a> and <a href="https://docs.docker.com/docker-for-windows/">Windows</a></p>

<p><strong>How to use CNV-Sim Docker image:</strong></p>

<div class="highlight highlight-source-shell"><pre>docker run -v <span class="pl-k">&lt;</span>absolute_local_path_to_input_directory<span class="pl-k">&gt;</span>:/my_data nabavilab/cnv-sim ./cnv-sim.py -o /my_data/<span class="pl-k">&lt;</span>simulation_name<span class="pl-k">&gt;</span> [OPTIONS] {genome, exome} /my_data/<span class="pl-k">&lt;</span>genome_file<span class="pl-k">&gt;</span> [/my_data/<span class="pl-k">&lt;</span>target_file<span class="pl-k">&gt;</span>]</pre></div>

<p>where:</p>

<ul>
<li>
<code>&lt;absolute_local_path_to_input_directory&gt;</code> is the directory where the required input files exist (genome.fa and target.bed if using exome simulation); <em>MUST</em> be absolute path </li>
<li>
<code>&lt;simulation_name&gt;</code> is used as the name of the output folder in your data directory, where the output of CNV Sim will be saved</li>
<li>
<code>&lt;genome_file&gt;</code> is the FASTA file name for the genome reference</li>
<li>
<code>&lt;target_file&gt;</code> is the BED file name for the targets (only if using exome as the simulation type)</li>
</ul>

<p>Refer to the below <a href="#cnv-sim-options">CNV-Sim options</a> section for more details on how to use all available command-line options or 
simply type: <code>docker run nabavilab/cnv-sim ./cnv-sim -h</code> to read the help.</p>

<h3>
<a id="use-from-galaxy" class="anchor" href="#use-from-galaxy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Use from Galaxy</h3>

<p>CNV-Sim is available as a <a href="https://galaxyproject.org/">Galaxy</a> tool. </p>

<p>Download it from the Galaxy Tool Shed: <a href="https://toolshed.g2.bx.psu.edu/view/ahosny/cnvsim/">https://toolshed.g2.bx.psu.edu/view/ahosny/cnvsim/</a></p>

<h2>
<a id="required-input" class="anchor" href="#required-input" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Required Input</h2>

<p><strong>IMPORTANT:</strong> for the current version of CNV-Sim, the underlying <em>pysam</em> library used Wessim might not be able to handle large reference files (such as the whole human genome) when introducing amplifications. To avoid errors, simulate CNV in one chromosome at a time (reference and target for one chromosome in a single run). We are working on handling large files seamlessly and will be released in the next version.</p>

<h3>
<a id="whole-genome" class="anchor" href="#whole-genome" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>whole genome</h3>

<ul>
<li>Reference genome in <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format.</li>
</ul>

<h3>
<a id="targeted-exome" class="anchor" href="#targeted-exome" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>targeted exome</h3>

<ul>
<li>Reference genome in <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format.</li>
<li>Target exons file in <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> format. 
The exome consists of introns and exons. The target file here should indicate the start and end positions of exons (not exomes).</li>
</ul>

<h2>
<a id="cnv-sim-options" class="anchor" href="#cnv-sim-options" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>CNV-Sim options</h2>

<pre><code>Usage: cnv-sim.py [-h] [-v] [-o OUTPUT_DIR_NAME] [-n N_READS] [-l READ_LENGTH]
                  [--cnv_list CNV_LIST] [-g REGIONS_COUNT]
                  [-r_min REGION_MINIMUM_LENGTH]
                  [-r_max REGION_MAXIMUM_LENGTH] [-a AMPLIFICATIONS]
                  [-d DELETIONS] [-cn_min COPY_NUMBER_MINIMUM]
                  [-cn_max COPY_NUMBER_MAXIMUM]
                  {genome,exome} genome [target]

Generates NGS short reads that encompass copy number variations in whole
genome and targeted exome sequencing

Positional arguments:
  {genome,exome}        simulate copy number variations in whole genome or
                        exome regions
  genome                path to the referece genome file in FASTA format
  target                path to the target regions file in BED format (if
                        using exome)

Optional arguments:
  -h, --help            show this help message and exit
  -v, --version         Show program's version number and exit.
  -o OUTPUT_DIR_NAME, --output_dir_name OUTPUT_DIR_NAME
                        a name to be used to create the output directory
                        (overrides existing directory with the same name).
  -n N_READS, --n_reads N_READS
                        total number of reads without variations
  -l READ_LENGTH, --read_length READ_LENGTH
                        read length (bp)
  --cnv_list CNV_LIST   path to a CNV list file in BED format chr | start |
                        end | variation. If not passed, it is randomly
                        generated using CNV list parameters below

CNV list parameters:
  parameters to be used if CNV list is not passed

  -g REGIONS_COUNT, --regions_count REGIONS_COUNT
                        number of CNV regions to be generated randomly
  -r_min REGION_MINIMUM_LENGTH, --region_minimum_length REGION_MINIMUM_LENGTH
                        minimum length of each CNV region
  -r_max REGION_MAXIMUM_LENGTH, --region_maximum_length REGION_MAXIMUM_LENGTH
                        maximum length of each CNV region
  -a AMPLIFICATIONS, --amplifications AMPLIFICATIONS
                        percentage of amplifications in range [0.0: 1.0].
  -d DELETIONS, --deletions DELETIONS
                        percentage of deletions in range [0.0: 1.0].
  -cn_min COPY_NUMBER_MINIMUM, --copy_number_minimum COPY_NUMBER_MINIMUM
                        minimum level of variations (copy number) introduced
  -cn_max COPY_NUMBER_MAXIMUM, --copy_number_maximum COPY_NUMBER_MAXIMUM
                        maximum level of variation (copy number) introduced
</code></pre>

<h2>
<a id="contributor" class="anchor" href="#contributor" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contributor</h2>

<p>Abdelrahman Hosny (<a href="https://github.com/abdelrahmanhosny" class="user-mention">@abdelrahmanhosny</a>)</p>

<h2>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>License</h2>

<p>The MIT License (MIT)</p>

<p>Copyright (c) 2016 <a href="https://nabavilab.github.io/">NabaviLab</a></p>

<h2>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Support or Contact</h2>

<p>Having trouble with running CNV-Sim? <a href="mailto:abdelrahman.hosny@ieee.org">contact us</a> and we’ll help you sort it out.</p>
        </section>

        <footer>
          CNV-Sim is maintained by <a href="https://github.com/NabaviLab">NabaviLab</a><br>
          This page was generated by <a href="https://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
